File: catkin_ws/src/localization/src/localization.cpp
#include <tf/tf.h>
#include "localization.h"
#include "UTM.cpp"
using namespace std;

// Constructor
Localization::Localization(){
    // Initialize
    m_orig_lat = 0.0;
    m_orig_lon = 0.0;
    m_orig_northing = 0.0;
    m_orig_easting = 0.0;
    m_zone = 51;

    // OnStartUp
    string key;
    if (ros::param::search("latitude", key)){
        ros::param::get(key, m_orig_lat); // Hz
	ROS_INFO("!!!!!!!!!!orig_lat : %f", m_orig_lat);
    }

    if (ros::param::search("longitude", key)){
        ros::param::get(key, m_orig_lon); // Hz
	ROS_INFO("!!!!!!!!!!orig_lon : %f", m_orig_lon);
        double lonInRad = DegToRad(m_orig_lon);
        double latInRad = DegToRad(m_orig_lat);
        MapLatLonToXY (latInRad, lonInRad, UTMCentralMeridian(m_zone), m_orig_easting, m_orig_northing);
        ROS_INFO("orig_easting    : %f", m_orig_easting);
        ROS_INFO("orig_northing    : %f",m_orig_northing);
   }

    // Publisher
    pub_NavX = m_nh.advertise<std_msgs::Float64>("NAV_X", 10);
    pub_NavY = m_nh.advertise<std_msgs::Float64>("NAV_Y", 10);
    pub_NavLon = m_nh.advertise<std_msgs::Float64>("NAV_LONG", 10);
    pub_NavLat = m_nh.advertise<std_msgs::Float64>("NAV_LAT", 10);
    pub_NavSpeed = m_nh.advertise<std_msgs::Float64>("NAV_SPEED_X", 10);
    pub_NavHdg = m_nh.advertise<std_msgs::Float64>("NAV_HEADING", 10);

    // Subscriber
    sub_gps = m_nh.subscribe("/fix", 1000, &Localization::callback_gps, this);
    sub_imu = m_nh.subscribe("/imu", 1000, &Localization::callback_imu, this);
}

// Destructor
Localization::~Localization(){
}

// Callback
void Localization::callback_gps(const sensor_msgs::NavSatFix::ConstPtr& msg){
    //Process gps lat lon
    double lat = msg->latitude;
    double lon = msg->longitude;
    double latInRad = DegToRad(lat);
    double lonInRad = DegToRad(lon);
    ROS_INFO("float lat  : %f", lat);
    ROS_INFO("Rad lat    : %f", latInRad);
    
    double easting = 0.0;
    double northing = 0.0;
    MapLatLonToXY (latInRad, lonInRad, UTMCentralMeridian(m_zone), easting, northing);
    //MapLatLonToXY (FLOAT lat in rad, FLOAT lon in rad , FLOAT zone in rad, FLOAT &x, FLOAT &y);
    ROS_INFO("easting    : %f", easting);
    ROS_INFO("northing    : %f", northing);

    // Calculate x,yi
    double x = easting - m_orig_easting;
    double y = northing - m_orig_northing;
    ROS_INFO("x    : %f", x);
    ROS_INFO("y    : %f", y);


    // publish to bridge topic
    pub_NavX.publish(toRosFloat(x));
    pub_NavY.publish(toRosFloat(y));
    pub_NavLat.publish(toRosFloat(lat));
    pub_NavLon.publish(toRosFloat(lon));
}

void Localization::callback_imu(const sensor_msgs::Imu msg){
    // transform quaterion into rpy
    tf::Quaternion q(msg.orientation.x,
                     msg.orientation.y,
                     msg.orientation.z,
                     msg.orientation.w);
    tf::Matrix3x3 m(q);
    double roll, pitch, yaw;
    m.getRPY(roll, pitch, yaw);
    ROS_INFO("yaw in rad    : %f", yaw);
    yaw = yaw*180.0/pi;
    ROS_INFO("yaw in degree : %f", yaw);
    pub_NavHdg.publish(toMoosAngle(yaw));
    pub_NavSpeed.publish(toRosFloat(0.0));
}

std_msgs::Float64 Localization::toRosFloat(double value){
    std_msgs::Float64 msg;
    msg.data = value;
    return msg;
}

std_msgs::Float64 Localization::toMoosAngle(double value){
    std_msgs::Float64 msg;
    if(value < 0.0)
        value = -value;
    else
	value = -value + 360.0;
    msg.data = value;
    return msg;
    //ROS_INFO("yaw to moos   : %f", value);
}

/*==============================*/
// Main
int main (int argc, char **argv)
{
    ros::init(argc, argv, "Localization_node");
    Localization node;
    ros::spin();
    return 0;
}

File: catkin_ws/src/localization/src/localization.py
#!/usr/bin/env python
'''
Author: Tony Hsiao                                                              
Date:2019/01/16                                                                
Last update: 2019/01/16                                                         
Locailization by gps and imu
'''
import rospy
from std_msgs.msg import Float64
from sensor_msgs.msg import NavSatFix, Imu
from diagnostic_msgs.msg import DiagnosticArray, DiagnosticStatus, KeyValue
from geodesy.utm import UTMPoint, fromLatLong
#import message_filters
from tf.transformations import euler_from_quaternion
import math
#from nav_msgs.msg import Odometry
#from geometry_msgs.msg import Pose
#import math
#from scipy.stats import norm
#import numpy as np

class Locailization(object):
    def __init__(self):
        
        # Params
        self.node_name = rospy.get_name()
        self.x = 0 # local coordinates
        self.y = 0

        # Onstartup
        self.lat_orig = rospy.get_param('~latitude', 0.0)  # (0,0) in moos
        self.long_orig = rospy.get_param('~longitude', 0.0)
        self.utm_orig = fromLatLong(self.lat_orig, self.long_orig)

        # Publisher
        self.pub_NavX = rospy.Publisher("NAV_X", Float64, queue_size=1);
        self.pub_NavY = rospy.Publisher("NAV_Y", Float64, queue_size=1);
        self.pub_NavLat = rospy.Publisher("NAV_LAT", Float64, queue_size=1);
        self.pub_NavLon = rospy.Publisher("NAV_LONG", Float64, queue_size=1);
        self.pub_NavHdg = rospy.Publisher("NAV_HEADING", Float64, queue_size=1);
        self.pub_NavSpeed = rospy.Publisher("NAV_SPEED_X", Float64, queue_size=1);

        #Subscriber
        self.imu_sub = rospy.Subscriber("/imu", Imu, self.cb_imu)
        self.gps_sub = rospy.Subscriber("/fix", NavSatFix, self.cb_gps)
        #sub_imu = message_filters.Subscriber("/imu", Imu)
        #sub_gps = message_filters.Subscriber("/fix", NavSatFix)
        #ats = message_filters.ApproximateTimeSynchronizer([sub_imu, sub_gps], queue_size = 5, slop = 1.0)
        #ats.registerCallback(self.cb_gps_imu)

    def cb_gps(self, msg):
        utm_point = fromLatLong(msg.latitude, msg.longitude)
        self.x = utm_point.easting - self.utm_orig.easting
        self.y = utm_point.northing - self.utm_orig.northing
        self.pub_NavX.publish(self.x)
        self.pub_NavY.publish(self.y)
        self.pub_NavLat.publish(msg.latitude)
        self.pub_NavLon.publish(msg.longitude)
        self.pub_NavSpeed.publish(0.0)


    def cb_imu(self, msg):
        quaternion = (
                msg.orientation.x,
                msg.orientation.y,
                msg.orientation.z,
                msg.orientation.w)

        (roll,pitch,yaw) = euler_from_quaternion(quaternion)
        self.pub_NavHdg.publish(yaw*(180.0/math.pi))
        #print(round(roll*(180.0/math.pi),2))

    #def cb_gps_imu(self, msg_imu, msg_gps):
    #    rospy.loginfo("cb_gps")
    #    self.cb_gps(msg_gps)
    #    self.cb_imu(msg_imu)


    def on_shutdown(self):
        rospy.loginfo("[%s] Shutdown." %(self.node_name))

if __name__ == '__main__':
    rospy.init_node('localization',anonymous=False)
    localization = Locailization()
    rospy.on_shutdown(localization.on_shutdown)
    rospy.spin()
